org-code-javabuilder/lib/src/main/java/org/code/javabuilder/UnhealthyContainerChecker.java

```diff
@@ -0,0 +1,58 @@
+package org.code.javabuilder;
+
+import com.amazonaws.services.dynamodbv2.AmazonDynamoDB;
+import com.amazonaws.services.dynamodbv2.model.AttributeValue;
+import java.util.Map;
+import org.code.protocol.LoggerUtils;
+
+/**
+ * Checks if the current container has been marked unhealthy, so that we can shut it down if needed.
+ */
+public class UnhealthyContainerChecker {
+  static final String CONTAINER_ID_KEY_NAME = "container_id";
+
+  /**
+   * When the health status is being checked. This allows us to choose whether to trigger a shutdown
+   * at the beginning or end of the session.
+   */
+  public enum ShutdownTrigger {
+    START("start"),
+    END("end");
+
+    private final String name;
+
+    ShutdownTrigger(String name) {
+      this.name = name;
+    }
+
+    public String getName() {
+      return this.name;
+    }
+  }
+
+  private final AmazonDynamoDB dynamoDBClient;
+  private final String tableName;
+
+  public UnhealthyContainerChecker(AmazonDynamoDB dynamoDBClient, String tableName) {
+    this.dynamoDBClient = dynamoDBClient;
+    this.tableName = tableName;
+  }
+
+  public boolean shouldForceShutdownContainer(String containerId, ShutdownTrigger trigger) {
+    // The container ID value is a concatenation of the ID and the shutdown trigger type
+    final String containerIdCompositeValue = containerId + "#" + trigger.getName();
+    final Map<String, AttributeValue> key =
+        Map.of(CONTAINER_ID_KEY_NAME, new AttributeValue(containerIdCompositeValue));
+    final Map<String, AttributeValue> entry;
+    try {
+      entry = this.dynamoDBClient.getItem(this.tableName, key).getItem();
```

Comment:
is there any risk of the dynamoDB client getting into a bad state (like the api gateway client sometimes does) and therefore throwing an exception here? I wonder if for now we should log a severe error instead of a warning so we know if this happens and can build in some checks so we don't cause user's programs to fail if the client is shut down.

Reply:
Hmm good question. I don't think there should be because it just queries the DB directly? But of course we have seen that new AWS-related errors do crop up from time to time so it would be good to be safe. For example I know if there is a permissions issue with the request, then this will throw. I can change this to a severe error. 
Re: "so we don't cause user's programs to fail if the client is shut down" - just to clarify, are you thinking there other checks we should make here? Currently any exception here is caught and we just return false so the container is not shut down, but will proceed normally otherwise.

Reply:
oh my bad I thought if false we would kill the program. In that case either warning or severe is fine. We probably should check our warnings on occasion though, I can update the new dashboard to try to give us a view into them

Reply:
I was mainly thinking of how sometimes aws was shutting down our api gateway client in terms of potential errors. The s3 client doesn't seem to have that issue though.

Reply:
Yeah agreed, I'd expect this to work more similarly to our S3 client, but it doesn't hurt to log the exception if we encounter it.